### ISSUE TITLE
Support telemetry correlation for applications running on 4.6 but compiled with 4.5 framework

### ISSUE THREAD

**Issue by SergeyKanzhelev** [dotnet/corefx#19052]
Diagnostics Source implements a workaround for the HttpClient requests monitoring by injecting code using reflection. This [logic](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs#L614-L651) only compiles for [NET46 library](https://github.com/dotnet/corefx/blob/b0665b6894a3cc0925217b7728bec54b5455dc2c/src/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj#L71-L73):

``` xml
<ItemGroup Condition="'$(TargetGroup)' == 'net46' OR '$(TargetGroup)' == 'netfx'">
  <Compile Include="System\Diagnostics\HttpHandlerDiagnosticListener.cs" />
</ItemGroup>
```

This means that customers compiling applications with the framework 4.5 and running an application as Azure Web App or on any other platforms with NET46 installed will not have correlation feature enabled.

Please compile this logic for NET45 and use a runtime check for the `System.HttpClient` library version. This will unblock the scenario.

/CC: @vancem, @lmolkova

**Comment by vancem** [comment]
@lmolkova - are you in a position to test V4.5 changes?

**Comment by SergeyKanzhelev** [comment]
Just to make sure we are on the same page - this issue is not for support of 4.5 framework. I propose the same reflection code will be running. The only difference is that detection of framework will happen in runtime, not compile time when you set the target framework for the project.

**Comment by lmolkova** [comment]
I've done some quick analysis on what could be done here

* I believe the right approach is to test it on NET45, most likely it will just work, so will will decide at compile time to include the hook.
I'll try to test it today, it's even a problem to have a machine with NET45 only.

*  Runtime detection.
1. System.Net.Http (or System.Net) dll version on NET45 and NET46 is the same: 4.0.0.0
2. `Environment.Version` only differs in Revision part and [this article](https://msdn.microsoft.com/en-us/library/hh925568(v=vs.110).aspx) guides that it is not correct to determine CLR version with `Environment.Version` and it should be determined based on registry. I do not think it's  a good idea to have this check in DiagnosticSource

**Comment by lmolkova** [comment]
I manually tested the hook on pure .NET45 with tests we have and it just works in the same way as for NET46. 
Here is PR: #19262

**Comment by lmolkova** [comment]
@SergeyKanzhelev 
Please note that this change will de-facto disable instrumentation with EventSource in ApplicationInsights

**Comment by SergeyKanzhelev** [comment]
@lmolkova it wouldn't disable it, it will improve it. My understanding is those callbacks are only for header insertion/reading

**Comment by vancem** [comment]
I don't understand @lmolkova 's comment above.   The change is basically treating V4.5 the same as V4.6, which, given that it does work on V4.5, seems to be only goodness.

**Comment by karelz** [comment]
Agreed with @vancem 
This looks scary, can you please explain it @lmolkova 

> Please note that this change will de-facto disable instrumentation with EventSource in ApplicationInsights?

**Comment by lmolkova** [comment]
@vancem @karelz 
Sorry for not being clear.

ApplicationInsights uses EventSource to track HTTP outgoing requests on NET45+.
Recently it started to listen to the HTTP hook implemented in DiagnosticSource.
The hook has higher priority than the EventSource since it gives more information (requests and responses).

So since the hook starts to work on NET45, EventSource listener becomes no-op. It is good and enables correlation features we are working on in AppInsights.

**Comment by karelz** [comment]
OK, thanks for explanation - so disabling EventSource is actually a GOOD thing, as it is replaced by usage of more detailed DiagnosticsSource.

**Comment by vancem** [comment]
Fixed with PR #19262
